server/plugins/client/bundle-handler-factory.js

```diff
@@ -0,0 +1,54 @@
+module.exports.createHandler = createHandler
+
+var fs = require('fs')
+
+var log = require('npmlog')
+var bundleClient = require('./bundle')
+
+function createHandler (hoodieClientPath, bundleTargetPath, bundleConfig) {
+  var bundlePromise
+
+  return function (request, reply) {
+    if (!bundlePromise) {
+      bundlePromise = new Promise(function (resolve, reject) {
+        log.silly('client', 'bundling')
+        bundleClient(hoodieClientPath, bundleTargetPath, bundleConfig, function (error, bundleBuffer, hasUpdate) {
+          if (error) {
+            return reject(error)
+          }
+
+          writeClientBundle(hasUpdate, bundleConfig.inMemory, bundleTargetPath, bundleBuffer)
+
+          resolve(bundleBuffer)
+        })
+      })
+    }
+
+    bundlePromise.then(function (buffer) {
+      reply(buffer).bytes(buffer.length).type('application/javascript')
+    }).catch(reply)
+  }
+}
```

Comment:
We usually don’t do factories, because they are harder to test. We usually prefer pre-binded methods (but that doesn’t mean that it’s always the right way, this is just a suggestion, up for discussion :)

```js
module.exports = createHandler
var bundlePromise

function createHandler (hoodieClientPath, bundleTargetPath, bundleConfig, request, reply) {
  if (!bundlePromise) {
    bundlePromise = new Promise(function (resolve, reject) {
      log.silly('client', 'bundling')
      bundleClient(hoodieClientPath, bundleTargetPath, bundleConfig, function (error, bundleBuffer, hasUpdate) {
        if (error) {
          return reject(error)
        }

        writeClientBundle(hasUpdate, bundleConfig.inMemory, bundleTargetPath, bundleBuffer)

        resolve(bundleBuffer)
      })
    })
  }

  bundlePromise.then(function (buffer) {
    reply(buffer).bytes(buffer.length).type('application/javascript')
  }).catch(reply)
}
```

And then in `server/plugins/client/index.js ` we could do 

```js
  server.route([{
    method: 'GET',
    path: '/hoodie/client.js',
    handler: createBundleHandler.bind(null, hoodieClientPath, bundleTargetPath, options.config)
  }])
```